Update: fix counting jsx comment len in max-len (fixes #12213)

yeonjuan opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
This pr will fix #12213.

If the comment is in the single line jsx comment, change the comment node for checking with JSXExpressionContainer. 

ex)

```jsx
<>
  {/* comment */}
</>
```


**Is there anything you'd like reviewers to focus on?**

kaicataldo commented:
CI is failing because of the issue reported [here](https://github.com/eslint/eslint/pull/12662).

yeonjuan commented:
@kaicataldo sure!

yeonjuan commented:
@kaicataldo Friendly ping :)

mdjermanovic commented on lib/rules/max-len.js:
```diff
@@ -183,6 +183,23 @@ module.exports = {
+
+            return parent.loc.start.line === parent.loc.end.line;
+        }
+
```
I'm not sure what would be correct behavior when there are two comments inside, like `{ /* 1 */ /* 2 */ }`.

A small difference is that this is now a `valid` comment-only line:

```js
{
    code: "var jsx = (<>\n" +
         "  {/* this line has > 20 characters */ /* but < 100 */}\n" +
         "</>)",
    options: [{ code: 20, comments: 100 }],
    parserOptions: { ecmaFeatures: { jsx: true } }
}
```

while this is `invalid`, because a line with two comments is treated as a 'code' line:

```js
{
    code: "var foo = \n" +
        "  /* this line has > 20 characters */ /* but < 100 */\n" +
        "5;",
     options: [{ code: 20, comments: 100 }],
     errors: 1 // "This line has a length of 53. Maximum allowed is 20."
}
```

Thoughts? Maybe this doesn't matter as much.

yeonjuan replied:
Although it needs to be evaluated by eslint members, In my opinion, the second case looks like a bug. :)

mdjermanovic replied:
I agree and would vote to do with JSX comments whatever seems to be easier for the other problem (with trailing comments).

yeonjuan commented on lib/rules/max-len.js:
```diff
@@ -305,6 +322,11 @@ module.exports = {
+                    // check jsx comment
+                    if (isCommentInSingleLineJSXEmptyExpression(comment)) {
+                        comment = sourceCode.getNodeByRangeIndex(comment.range[0]).parent;
+                    }
```
> This might execute relatively costly isCommentInSingleLineJSXEmptyExpression on some irrelevant comments like those inside code, but could end up being more optimal because the do-while loop in lines.forEach often gets back to the same comment, if I correctly understand the loop?

Thanks for the always careful review! I agree and, I'll care of it.

yeonjuan commented:
@kaicataldo 
Thanks!. I added more test cases you mentioned. 
But Maybe there would be no change in those cases compared with the actual version.

@mdjermanovic 
I changed the code and commit.  Thanks! :)

mdjermanovic commented on tests/lib/rules/max-len.js:
```diff
@@ -650,6 +734,204 @@ ruleTester.run("max-len", rule, {
+            code: "var jsx = (<>\n" +
+                  "{ 38/* this line has 38 characters */}\n" +
+                  "</>)",
+            options: [15, { comments: 37 }],
```
This could be `comments: 38` to show that line fails because it isn't a comment?

Also, it feels like some tests with an expression (`a && b`) + comment inside `{}` might be missing `ignoreTrailingComments: true` or `ignoreComments: true` to show that this doesn't count as a comment?

And I think (could be wrong) that there is no test where there is only a comment inside `{}`, but `{` and `}` are not on the same line so it doesn't count as a comment because of that.

yeonjuan replied:
Sure!

yeonjuan replied:
@mdjermanovic 
Humm.. what do you think about this case? A multi-line comment in JSXEmptyExpression should be treated as a code?
```jsx
var jsx = (<foo>
     {/* multiline comment
         without code */ }
</foo>);
```
so if then, Should I treat the case below as an invalid?
```
 {
     code: "var jsx = (<>\n" +
                  "  {/*  multi-line          comment\n" +
                  "                   but it is a code ? */ }\n"
                  "</>)",
    options: [20, { ignoreComments: true }],
    parserOptions: { ecmaFeatures: { jsx: true } }
 }
```

yeonjuan replied:
@mdjermanovic 
> I think the decision to treat the whole { /* */ } as a comment only if everything incl. braces is on the same line is a good decision, since it's a minimal change in behavior.

Humm, I'm confusing 🤔
Maybe treating the whole {/* ... *} as a comment in this PR would be better? ( and change tag `fix` to `update`)

mdjermanovic replied:
Sorry for the confusion! Just wanted to say that I agree that only those `{/* */}` where everything is on the same line should be treated as comments, as you already implemented with the `parent.loc.start.line === parent.loc.end.line` condition :+1:

yeonjuan commented on lib/rules/max-len.js:
```diff
@@ -183,6 +183,21 @@ module.exports = {
+
+            const parent = node.parent;
+
+            return parent.loc.start.line === parent.loc.end.line;
```
@mdjermanovic 
Thanks. you are definitely right. :)

kaicataldo commented:
I think it's fine to accept this as a semver-minor bug fix (though, in practice I'm not sure it matters since we are merging breaking changes in preparation for v8 right now).

kaicataldo commented:
Thanks for contributing to ESLint!
